include/ensmallen_bits/ftml/ftml.hpp

```diff
@@ -0,0 +1,159 @@
+/**
+ * @file ftml.hpp
+ * @author Marcus Edel
+ *
+ * Definition of Follow the Moving Leader (FTML).
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_FTML_FTML_HPP
+#define ENSMALLEN_FTML_FTML_HPP
+
+#include <ensmallen_bits/sgd/sgd.hpp>
+
+#include "ftml_update.hpp"
+
+namespace ens {
+
+/**
+ * Follow the Moving Leader (FTML) is an optimizer where recent samples are
+ * weighted more heavily in each iteration, so FTML can adapt more quickly to
+ * changes.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @inproceedings{Zheng2017,
+ *   author    = {Shuai Zheng and James T. Kwok},
+ *   title     = {Follow the Moving Leader in Deep Learning},
+ *   year      = {2017}
+ *   booktitle = {Proceedings of the 34th International Conference on Machine
+ *                Learning},
+ *   pages     = {4110--4119},
+ *   series    = {Proceedings of Machine Learning Research},
+ *   publisher = {PMLR},
+ * }
+ * @endcode
+ *
+ * For FTML to work, a DecomposableFunctionType template parameter is
+ * required. This class must implement the following function:
```

Comment:
We have this documentation across a lot of optimizers, but the addition of `function_types.md` probably means we can reduce and centralize a lot of it.  Do you think that we could replace this with something like... `FTML requires a separable differentiable function to optimize (see <url>).`  The only problem with that is that it's not clear what URL to use there.  I suppose ideally we'd like to point a user at the website documentation, but, the URL there could change.  So I am not sure what the best choice is.

Reply:
Actually for this one I wonder if it would be best to open another issue instead of handling it in this PR.